spot-vaults/contracts/_strategies/SpotPricer.sol

```diff
@@ -0,0 +1,295 @@
+// solhint-disable-next-line compiler-version
+pragma solidity ^0.7.6;
+pragma abicoder v2;
+
+import { FullMath } from "@uniswap/v3-core/contracts/libraries/FullMath.sol";
+import { UniswapV3PoolHelpers } from "../_utils/UniswapV3PoolHelpers.sol";
+
+import { IUniswapV3Pool } from "@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol";
+import { IERC20 } from "../_interfaces/external/IERC20.sol";
+import { IWAMPL } from "../_interfaces/external/IWAMPL.sol";
+import { IPerpetualTranche } from "../_interfaces/external/IPerpetualTranche.sol";
+import { IChainlinkOracle } from "../_interfaces/external/IChainlinkOracle.sol";
+import { IAmpleforthOracle } from "../_interfaces/external/IAmpleforthOracle.sol";
+
+import { IPerpPricer } from "../_interfaces/IPerpPricer.sol";
+import { IMetaOracle } from "../_interfaces/IMetaOracle.sol";
+
+/**
+ * @title SpotPricer
+ *
+ * @notice A pricing oracle for SPOT, a perpetual claim on AMPL senior tranches.
+ *
+ *         Internally aggregates prices from multiple oracles.
+ *         Chainlink for USDC and ETH prices,
+ *         The Ampleforth CPI oracle for the AMPL price target and
+ *         UniV3 pools for current AMPL and SPOT market prices.
+ *
+ */
+contract SpotPricer is IPerpPricer, IMetaOracle {
+    //-------------------------------------------------------------------------
+    // Constants
+
+    /// @dev Standardizes prices from various oracles and returns the final value
+    ///      as a fixed point number with {DECIMALS} places.
+    uint256 private constant DECIMALS = 18;
+    uint256 private constant ONE = (10 ** DECIMALS);
+
+    /// @dev We bound the deviation factor to 100.0.
+    uint256 public constant MAX_DEVIATION = 100 * ONE; // 100.0
+
+    /// @dev Token denominations.
+    uint256 private constant ONE_USDC = 1e6;
+    uint256 private constant ONE_WETH = 1e18;
+    uint256 private constant ONE_SPOT = 1e9;
+    uint256 private constant ONE_AMPL = 1e9;
+    uint256 private constant ONE_WAMPL = 1e18;
+
+    /// @dev Oracle constants.
+    uint256 private constant CL_ETH_ORACLE_STALENESS_THRESHOLD_SEC = 3600 * 12; // 12 hours
+    uint256 private constant CL_USDC_ORACLE_STALENESS_THRESHOLD_SEC = 3600 * 48; // 2 day
+    uint256 private constant USDC_UPPER_BOUND = (101 * ONE) / 100; // 1.01$
+    uint256 private constant USDC_LOWER_BOUND = (99 * ONE) / 100; // 0.99$
+    uint32 private constant UNIV3_TWAP_DURATION = 3600;
+
+    //--------------------------------------------------------------------------
+    // Modifiers
+
+    /// @dev Throws if called by any account other than the owner.
+    modifier onlyOwner() {
+        // solhint-disable-next-line custom-errors
+        require(msg.sender == owner, "UnauthorizedCall");
+        _;
+    }
+
+    //-------------------------------------------------------------------------
+    // Storage
+
+    /// @notice Address of the WETH-WAMPL univ3 pool.
+    IUniswapV3Pool public immutable WETH_WAMPL_POOL;
```

Comment:
Thinking about how this might be used by downstream clients... this layout is very specific to our current liquidity landscape.

If UniV4 gets released, or even if wampl liquidity moved from a 1% pool to a .3% pool, then we'd need to redeploy a new SpotPricer contract, and and the users would need to update their callsites.

We could make this more configurable, or we could deploy a router layer on top of pool-specific contracts. Not super excited to deploy a whole new call layer. Maybe we just advise clients to make their IMetaOracle references configurable